WooCommerce/Classes/ViewRelated/Products/Edit Product/Downloads/Downloadable FIle List/ProductDownloadListViewModel.swift

```diff
@@ -0,0 +1,151 @@
+import Foundation
+import Yosemite
+
+/// Provides data needed for downloadable files settings.
+///
+protocol ProductDownloadListViewModelOutput {
+    var downloads: [ProductDownloadDnD] { get }
+    var downloadLimit: Int64? { get }
+    var downloadExpiry: Int64? { get }
+
+        // Convenience Methodes
+    @discardableResult
+    func remove(at index: Int) -> ProductDownloadDnD?
+    @discardableResult
+    func item(at index: Int) -> ProductDownloadDnD?
+    func insert(_ newElement: ProductDownloadDnD, at index: Int)
+    func count() -> Int
+}
+
+/// Handles actions related to the downloadable files settings data.
+///
+protocol ProductDownloadListActionHandler {
+    // Tap actions
+    func didTapDownloadableFileFromRow(_ indexPath: IndexPath)
+
+    // Input field actions
+    func handleDownloadsChange(_ downloads: [ProductDownload])
+    func handleDownloadLimitChange(_ downloadLimit: Int64?)
+    func handleDownloadExpiryChange(_ downloadExpiry: Int64?)
+
+    // Navigation actions
+    func completeUpdating(onCompletion: ProductDownloadListViewController.Completion)
+    func hasUnsavedChanges() -> Bool
+}
+
+protocol ProductDownloadListDataSource {
+
+}
+
+/// Error cases that could occur in product downloadable files settings.
+///
+enum ProductDownloadListError: Error {
+    case emptyFileName
+    case emptyUrl
+}
+
+/// Provides view data for price settings, and handles init/UI/navigation actions needed in product price settings.
+///
+final class ProductDownloadListViewModel: ProductDownloadListViewModelOutput {
+    private let product: ProductFormDataModel
+
+    // Editable data
+    //
+    private(set) var downloads = [ProductDownloadDnD]()
+    private(set) var downloadLimit: Int64?
+    private(set) var downloadExpiry: Int64?
+
+    init(product: ProductFormDataModel) {
+        self.product = product
+
+        downloads = populateDataSource(product.downloads)
+        downloadLimit = product.downloadLimit
+        downloadExpiry = product.downloadExpiry
+    }
+
+    // MARK: - ProductDownloadListDataSource Methodes
+    //
+    func remove(at index: Int) -> ProductDownloadDnD? {
+        return downloads.remove(at: index)
+    }
+
+    func item(at index: Int) -> ProductDownloadDnD? {
+        return downloads[index]
+    }
+
+    func insert(_ newElement: ProductDownloadDnD, at index: Int) {
+        downloads.insert(newElement, at: index)
+    }
+
+    func count() -> Int {
+        return downloads.count
+    }
+
+    func allDownloadableFiles(_ downloadableFiles: [ProductDownloadDnD]) -> [ProductDownload] {
+        var downloads = [ProductDownload]()
+        for download in downloadableFiles {
+            downloads.append(download.download)
+        }
+        return downloads
```

Comment:
This can be optimized and it's probably not needed a method to embed it
```suggestion
        return downloadableFiles.map { $0.download }
```

Reply:
What was I thinking! How did I miss that!  😢